pkg/controllers/user/rbac/namespace_handler.go

```diff
@@ -453,11 +460,97 @@ func updateStatusAnnotation(hasPRTBs bool, namespace *v1.Namespace, mgr *manager
 			logrus.Warnf("fail to set %v condition on ns %v: %v", initialRoleCondition, namespace.Name, err)
 			continue
 		}
-		_, err = mgr.workload.Core.Namespaces("").Update(ns)
+		_, err = mgr.Workload.Core.Namespaces("").Update(ns)
 		if err == nil {
 			break
 		}
 		logrus.Warnf("error updating ns %v status: %v", ns.Name, err)
 	}
 
 }
+
+func (n *nsLifecycle) ensureAppRolesAddToNamespaces(ns *v1.Namespace) (bool, error) {
+	// Get project that contain this namespace
+	projectID := ns.Annotations[projectIDAnnotation]
+	if len(projectID) == 0 {
+		rbs, err := n.m.RBLister.List(ns.Name, labels.Everything())
+		if err != nil {
+			return false, errors.Wrapf(err, "couldn't list role bindings in %s", ns.Name)
+		}
+		client := n.m.Workload.RBAC.RoleBindings(ns.Name).ObjectClient()
+		for _, rb := range rbs {
+			if uid := convert.ToString(rb.Labels[RTBOwnerLabel]); uid != "" {
+				logrus.Infof("Deleting role binding %s in %s", rb.Name, ns.Name)
+				if err := client.Delete(rb.Name, &metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+					return false, errors.Wrapf(err, "couldn't delete role binding %s", rb.Name)
+				}
+			}
+		}
+		return false, nil
+	}
+	_, projectName := ref.Parse(projectID)
+
+	// prtbs, err := n.m.prtbIndexer.ByIndex(prtbByProjectIndex, projectID)
+	apps, err := n.m.AppLister.List(projectName, labels.Everything())
+	if err != nil {
+		return false, errors.Wrapf(err, "couldn't get project apps in project id %s", projectID)
+	}
+	hasApps := len(apps) > 0
+
+	for _, app := range apps {
+		roleTemplateName := helmcommon.GetInjectRoleTemplateName(app)
+
+		if app.DeletionTimestamp != nil || roleTemplateName == "" {
+			continue
+		}
+
+		rt, err := n.m.RTLister.Get("", roleTemplateName)
```

Comment:
this method is a bit too long. Would help if some parts of it can be extracted. Also you should add more comments on what this method does.